src/backend/distributed/executor/multi_utility.c

```diff
@@ -1378,41 +1631,85 @@ ShowNoticeIfNotUsing2PC(void)
 /*
  * DDLCommandList builds a list of tasks to execute a DDL command on a
  * given list of shards.
+ *
+ * leftRelationId is the relation id of actual distributed table which given DDL command
+ * is applied. rightRelationId is used if there is another relation which is affected by
+ * executed DDL command. At the moment it is only used for passing referenced table id
+ * around for ALTER TABLE ... ADD CONSTRAINT FOREIGN KEY queries.
  */
 static List *
-DDLTaskList(Oid relationId, const char *commandString)
+DDLTaskList(Oid leftRelationId, Oid rightRelationId, const char *commandString)
```

Comment:
Could you split this function as well as ExecuteDistributedDDLCommand into separate functions (e.g. DDLTaskList, AddForeignKeyTaskList)? That would make it more readable and then you can just use forboth(leftShardCell, leftShardList, rightShardCell, rightShardList) to iterate through both lists.